crates/core/tedge_api/src/address.rs

```diff
@@ -0,0 +1,220 @@
+    _pd: PhantomData<fn(MB)>,
+    sender: MessageSender,
+}
+
+impl<MB: ReceiverBundle> std::fmt::Debug for Address<MB> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct(&format!("Address<{}>", std::any::type_name::<MB>()))
+            .finish_non_exhaustive()
+    }
+}
+
+impl<MB: ReceiverBundle> Clone for Address<MB> {
+    fn clone(&self) -> Self {
+        Self {
+            _pd: PhantomData,
+            sender: self.sender.clone(),
+        }
+    }
+}
+
+impl<MB: ReceiverBundle> Address<MB> {
+    /// THIS IS NOT PART OF THE PUBLIC API, AND MAY CHANGE AT ANY TIME
+    #[doc(hidden)]
+    pub fn new(sender: MessageSender) -> Self {
+        Self {
+            _pd: PhantomData,
+            sender,
+        }
+    }
+
+    /// Send a message `M` to the address represented by the instance of this struct
+    ///
+    /// This function can be used to send a message of type `M` to the plugin that is addressed by
+    /// the instance of this type.
+    ///
+    /// # Return
+    ///
+    /// The function either returns `Ok(())` if sending the message succeeded,
+    /// or the message in the error variant of the `Result`: `Err(M)`.
+    ///
+    /// The error is returned if the receiving side (the plugin that is addressed) does not receive
+    /// messages anymore.
+    ///
+    /// # Details
+    ///
+    /// For details on sending and receiving, see `tokio::sync::mpsc::Sender`.
+    pub async fn send<M: Message>(&self, msg: M) -> Result<ReplyReceiver<M::Reply>, M>
+    where
+        MB: Contains<M>,
+    {
+        let (sender, receiver) = tokio::sync::oneshot::channel();
+
+        self.sender
+            .send(InternalMessage {
+                data: Box::new(msg),
+                reply_sender: sender,
+            })
+            .await
+            .map_err(|msg| *msg.0.data.downcast::<M>().unwrap())?;
+
+        Ok(ReplyReceiver {
+            _pd: PhantomData,
+            reply_recv: receiver,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct ReplyReceiver<M> {
+    _pd: PhantomData<fn(M)>,
+    reply_recv: tokio::sync::oneshot::Receiver<AnySendBox>,
+}
+
+impl<M: Message> ReplyReceiver<M> {
+    pub async fn wait_for_reply(self, timeout: Duration) -> Result<M, ReplyError> {
+        let data = tokio::time::timeout(timeout, self.reply_recv)
+            .await
+            .map_err(|_| ReplyError::Timeout)?
+            .map_err(|_| ReplyError::Unknown)?;
+
+        Ok(*data.downcast().expect("Invalid type received"))
+    }
+}
+
+#[derive(Debug)]
+pub struct ReplySender<M> {
+    _pd: PhantomData<fn(M)>,
+    reply_sender: tokio::sync::oneshot::Sender<AnySendBox>,
+}
+
+impl<M: Message> ReplySender<M> {
+    pub(crate) fn new(reply_sender: tokio::sync::oneshot::Sender<AnySendBox>) -> Self {
+        Self {
+            _pd: PhantomData,
+            reply_sender,
+        }
+    }
+
+    pub fn reply(self, msg: M) -> Result<(), M> {
```

Comment:
Right now you cannot wait for a reply on a reply... This might be cool/useful. So should be added.